### PR TITLE
Perception

### PR DESCRIPTION
**Pull request by nathanestill**
We made colortracking publish the roomba's pose to tracker/debug.

### CODE REVIEW

File: iarc_vision/scripts/colortracking.py [hunk 1]
```diff
@@ -1,60 +1,298 @@
...
+                heading = get_heading(box, center, self.processed_image)
+                ray = self.cameraModel.projectPixelTo3dRay(center)
+                camera_ray = Vector3Stamped(header=msg.header,
+                                            vector=Vector3(*ray))
+                world_ray = self.tf.transformVector3('map', camera_ray)
+                # print(camera_ray, world_ray)
+                pos, quat = self.tf.lookupTransform('map', msg.header.frame_id, msg.header.stamp)
+                multiplier = -pos[2] / world_ray.vector.z
+                drone_to_roomba = np.array([world_ray.vector.x, world_ray.vector.y, world_ray.vector.z]) * multiplier
+                quat = quaternion_from_euler(0,0,heading)
```
Comment (HALtheWise) [inline iarc_vision/scripts/colortracking.py]: Does this also need to account for the orientation of the drone in the world that was discovered above?

File: iarc_vision/scripts/colortracking.py [hunk 2]
```diff
@@ -1,60 +1,298 @@
...
+
+            for box in self.boxes:
+                center = np.mean(box, axis=0)
+                heading = get_heading(box, center, self.processed_image)
+                ray = self.cameraModel.projectPixelTo3dRay(center)
+                camera_ray = Vector3Stamped(header=msg.header,
+                                            vector=Vector3(*ray))
+                world_ray = self.tf.transformVector3('map', camera_ray)
+                # print(camera_ray, world_ray)
+                pos, quat = self.tf.lookupTransform('map', msg.header.frame_id, msg.header.stamp)
```
Comment (HALtheWise) [inline iarc_vision/scripts/colortracking.py]: `quat` looks like it is never used before being overwritten below. We should either use it or replace it with `_`

File: iarc_vision/scripts/colortracking.py [hunk 3]
```diff
@@ -1,60 +1,298 @@
...
+                quat = quaternion_from_euler(0,0,heading)
+                map_to_roomba = pos + drone_to_roomba
+                print map_to_roomba
+
+                pose = Pose(position=Point(*map_to_roomba), orientation = Quaternion(*quat))
+
+                pwcs = PoseWithCovarianceStamped(header=Header(frame_id='map', stamp=msg.header.stamp),
+                                                 pose=PoseWithCovariance(
+                                                     pose=pose,
+                                                     covariance=np.diag([.2, .2, 0, 0, 0, 0.2]).flatten()
```
Comment (HALtheWise) [inline iarc_vision/scripts/colortracking.py]: This .2 meter and 0.2 radian uncertainty could probably be a constant somewhere, or even a ROS parameter if we want to get fancy.

**Review by HALtheWise — APPROVED** [review]
This looks fine, I left a couple comments in case you want to make a few more tweaks, but I'm fine with this getting merged as-is.

**Comment by yycho0108** [comment]
Please wait until I can also review these changes for integration compatibility.

**Comment by yycho0108** [comment]
The color tracker changes have been integrated into the perception pipeline. The pull request into master would require a separate review for the perception.

In terms of colortracker-exclusive comments, see notes under `colortracker.py` for the necessary upgrades.

**Pull request closed without merging**